### DISCUSSION TITLE
Unexpected lock contention in a sqlalchemy queuepool while using ~200 simultaneous threads

### DISCUSSION THREAD

**Discussion by gibsondan** [sqlalchemy/sqlalchemy#11102] [Usage Questions]
Hello, I'm seeing unexpected results when trying to use SqlAlchemy connection pooling using sqlalchemy 1.4.51, and I'm curious if this is a known issue.

I have approximately 200 threads that are making frequent database queries, using a standard sqlalchemy QueuePool - max_pool of 32 with 512 overflow in our latest set of experiments.

We've found that as we increase the throughput of the system by removing other bottlenecks besides our database queries, we are suddenly reliably able to get into a state that looks very similar to a deadlock, in that progress essentially crawls completely to a halt. Running `py-spy dump` on the system when everything is stuck reveals the following:

- About 100 threads are stuck in a callsite like this while attempting to put back their database connection, waiting for a condition variable: https://github.com/sqlalchemy/sqlalchemy/blob/rel_1_4_51/lib/sqlalchemy/util/queue.py#L104
```
    __enter__ (threading.py:272)
    put (sqlalchemy/util/queue.py:104)
    _do_return_conn (sqlalchemy/pool/impl.py:112)
    _return_conn (sqlalchemy/pool/base.py:336)
    checkin (sqlalchemy/pool/base.py:546)
    _finalize_fairy (sqlalchemy/pool/base.py:800)
    _checkin (sqlalchemy/pool/base.py:1008)
    close (sqlalchemy/pool/base.py:1166)
    close (sqlalchemy/engine/base.py:1251)
    __exit__ (sqlalchemy/engine/base.py:219)
```

- The other 100 threads are stuck in this callsite while attempting to get a new connection, waiting for a different condition variable backed by the same lock: https://github.com/sqlalchemy/sqlalchemy/blob/rel_1_4_51/lib/sqlalchemy/util/queue.py#L143

```
    __enter__ (threading.py:272)
    get (sqlalchemy/util/queue.py:143)
    _do_get (sqlalchemy/pool/impl.py:124)
    checkout (sqlalchemy/pool/base.py:493)
    _checkout (sqlalchemy/pool/base.py:894)
    connect (sqlalchemy/pool/base.py:327)
    _wrap_pool_connect (sqlalchemy/engine/base.py:3371)
    raw_connection (sqlalchemy/engine/base.py:3404)
    __init__ (sqlalchemy/engine/base.py:96)
    connect (sqlalchemy/engine/base.py:3325)
```

- no thread is actually doing any database work, and no thread has a stack trace that indicates that it is the one that is currently holding the lock that everybody is waiting for. Every thread is either about to start a db query and waiting to get a connection, or wrapping up a db query and trying to return their connection to the pool. 

- There are still plenty of available connections (we are including 512 overflow connections, well over the total number of threads that might be asking for them), so the issue isn't that there aren't enough connections, but rather this strange stuck state where no work is getting done.

I'm happy to provide any additional information that might help understand what's going on here, but I'm curious if we are hitting any known issues or bottlenecks in the connection pooling system. Thank you very much for any help you can provide.

**Comment by gibsondan** [comment]
Another data point here is that we don't seem to get into this strange deadlock-like state (or at least, it seems to take longer to reach it) if we reduce the number of threads concurrently accessing the queue (to, say, ~64 instead of ~200).

**Comment by zzzeek** [comment]
hi -

no particular insight on this, sorry, my experience with cPython has always been using more than 50 threads will generally start to fall over.   You may observe the queue that's in use here is basically a vendored version of Python's own [queue](https://docs.python.org/3/library/queue.html), so you would in theory be able to simulate the same locking occurring with just a plain `Queue` putting and getting objects.   The one thing we change about `Queue` is that we are using a `threading.RLock` instead of a `Lock`, which is to provide for garbage collected connections being returned to the pool in unexpected places.

**Reply by zzzeek** [reply]
also, this is code that's worked the same way more or less for about 18 years, which is not to say there is not a systemic issue in the pool implementation, but that while we've had **lots** of pool issues reported, I've not seen this one.

you might want to look into the [use_lifo](https://docs.sqlalchemy.org/en/20/core/pooling.html#pool-use-lifo) parameter that was added some years back by a contributor, which reverses the queing scheme to behave more like a stack.  not sure if that changes anything here.

**Reply by gibsondan** [reply]
Got it - thank you for the confirmation that this is not expected / the symptoms are unusual.

**Reply by gibsondan** [reply]
I was also looking through https://github.com/sqlalchemy/sqlalchemy/issues/7679 which doesn't seem exactly the same, but has some thematic overlap - there they were having only some threads getting starved, unlike our situation where somehow *all* (or almost all) threads seem to be starved. But it's possible that the "sprinkle in some sleeps" solution could help us out here too.

**Reply by zzzeek** [reply]
wow yes, that is totally likely related to what you have if not the same thing.     so it's come up before, I'd look at the guidance on that issue.

**Reply by gibsondan** [reply]
Based on that issue I tried using this simple pool implementation that added a sleep before each get or put - it may have helped delay the bad state, but we still got into the bad state after about 10 minutes sadly.

```
class SleepingQueuePool(db.QueuePool):
    """QueuePool that sleeps before each get or put, in an attempt to reduce
    thread starvation when a large number of db connections are happening in parallel
    in different threads.
    """

    def _do_get(self):
        time.sleep(1e-9)
        return super()._do_get()

    def _do_return_conn(self, conn):
        time.sleep(1e-9)
        return super()._do_return_conn(conn)
```

**Reply by zzzeek** [reply]
this is the kind of thing where cPython version and OS can be significant.  maybe try a different Python version?

**Comment by gibsondan** [comment]
@zzzeek I know this thread is quite old but I realized I never followed up here with how we got ourselves unblocked from this issue, in case it is interesting to you or potentially useful to others in a similar situation. In certain situations where we were hitting this problem, we started using our own QueuePool that doesn't use locks:

```py
import sqlalchemy.pool as db_pool

LOCK_FREE_QUEUE_POOL_SLEEP_INVERVAL_SECONDS = 0.1

class LockFreeQueuePool(db_pool.Pool):
    """Lock-free pool inspired by the sqlalchemy QueuePool:
    https://github.com/sqlalchemy/sqlalchemy/blob/rel_1_4_51/lib/sqlalchemy/pool/impl.py#L27.
    """

    _is_asyncio: bool = False  # pyright: ignore[reportIncompatibleVariableOverride]

    def __init__(
        self,
        creator,
        pool_size,
        max_overflow,
        timeout,
        **kw,
    ):
        db_pool.Pool.__init__(self, creator, **kw)
        self._pool = []
        self._checked_out = set()
        self._max_pool_size = pool_size
        self._max_overflow = max_overflow
        self._timeout = timeout

    def _do_return_conn(self, record: db_pool.ConnectionPoolEntry):
        try:
            self._checked_out.remove(id(record))
        except KeyError:
            self.logger.warning("Connection not found in checked out set")

        if len(self._pool) >= self._max_pool_size:
            # returned conn is an overflow, dont save
            record.close()
        else:
            self._pool.append(record)

    def _do_get(self):
        timeout = self._timeout
        while True:
            try:
                conn = self._pool.pop(0)  # pop from front of list, FIFO
                self._checked_out.add(id(conn))
                return conn
            except IndexError:
                pass

            if len(self._checked_out) <= self._max_pool_size + self._max_overflow:
                # this is slightly racy - could result in dipping above max_overflow
                conn = self._create_connection()
                self._checked_out.add(id(conn))
                return conn

            if timeout <= 0:
                raise db.exc.TimeoutError(
                    "Outstanding connection count %s reached - no more connections for now"
                    % (self._max_pool_size + self._max_overflow)
                )

            timeout = timeout - LOCK_FREE_QUEUE_POOL_SLEEP_INVERVAL_SECONDS
            time.sleep(LOCK_FREE_QUEUE_POOL_SLEEP_INVERVAL_SECONDS)

    def recreate(self):
        self.logger.info("Pool recreating")
        return self.__class__(
            self._creator,
            pool_size=self._max_pool_size,
            max_overflow=self._max_overflow,
            timeout=self._timeout,
            pre_ping=self._pre_ping,
            recycle=self._recycle,
            echo=self.echo,
            logging_name=self._orig_logging_name,
            reset_on_return=self._reset_on_return,
            _dispatch=self.dispatch,
            dialect=self._dialect,
        )

    def dispose(self):
        while True:
            try:
                conn = self._pool.pop()
                conn.close()
            except IndexError:
                break

        self.logger.info("Pool disposed. %s", self.status())

    def status(self):
        return (
            "Lock-free pool size: %d  Connections in pool: %d "  # noqa: UP031
            "Current Overflow: %d Current Checked out "
            "connections: %d"
            % (
                self.size(),
                self.checkedin(),
                self.overflow(),
                self.checkedout(),
            )
        )

    def size(self):
        return self._max_pool_size

    def checkedin(self):
        return len(self._pool)

    def overflow(self):
        # Value is negative when no overflow is being used
        return len(self._checked_out) + len(self._pool) - self._max_pool_size

    def checkedout(self):
        return len(self._checked_out)
```

This has some obvious downsides (in particular, that its now possible for it to exceed its overflow) but does work around the starvation issue.

Belated thanks for helping us understand the problem.

**Reply by gibsondan** [reply]
Couple other details:
- For whatever reason, the 'just add a sleep' solution in the linked thread didn't seem to really help for us, we were still able to enter the deadlock state pretty reliably.
- We didn't do an exhaustive python version search, but the starvation issue was still occurring as of Python 3.12.11.

**Reply by zzzeek** [reply]
id love to know if the issue is specific to queue.queue or the _overflow_lock.   since you have a reproducer and willingness to hack on QueuePool would you want to try swapping out one or the other and seeing which it is? (or if it's both at the same time, which might not be that surprising if they are in some kind of mutually-exclusion situation)
